File: app/src/lib/pipeline.ts
// Pipeline client — wraps POST /predict on the FastAPI service.
// Falls back to a local synthesiser using key_numbers_python.json + the real
// pipeline_meta.json (synced from serve/models/meta.json by `npm run sync-data`)
// when the API is unreachable or VITE_PIPELINE_API is unset, so the screen still
// renders honestly in offline / pre-deploy demos.

import canon from '../data/key_numbers_python.json';
import metaFile from '../data/pipeline_meta.json';

export type Mode = 'hindcast' | 'forward';

export interface PredictRequest {
  mode: Mode;
  country: string;
  scenario?: string;
  elasticity?: number;
  gwp_usdm?: number;
  base_lr?: number;
  target_year?: number;
  feature_overrides?: Partial<Record<string, number>>;
}

export interface Trace {
  stage_1_inputs: {
    country: string;
    year: number;
    mode: Mode;
    raw_features: Record<string, number>;
    applied_overrides: Record<string, number>;
  };
  stage_2_features: {
    feature_order: string[];
    X: number[];
    log_transformed_keys: string[];
  };
  stage_3_xgb: {
    log_ghg_pred: number;
    ghg_pred_Mt: number;
    actual_Mt: number | null;
    err_pct: number | null;
    inference_ms: number | null;
    model: 'M3a' | 'M3b';
  };
  stage_4_scenario: {
    scenario: string;
    growth_rate_pa: number;
    years_compounded: number;
    hothouse_Mt: number;
    target_Mt: number;
    delta_pct: number;
  };
  stage_5_loss: {
    lr: number;
    lr_pp_vs_base: number;
    loss_USDm: number;
    loss_swing_vs_hothouse_USDm: number;
  };
  trace_meta: {
    pipeline_version: string;
    seed: number;
    total_latency_ms: number | null;
    served_by: 'fastapi' | 'cached';
    served_at: string;
  };
}

export interface PipelineMeta {
  pipeline_version: string;
  random_state: number;
  m3a_features: string[];
  m3b_features: string[];
  countries: string[];
  last_actual_year: number;
  ngfs_scenarios: Record<string, number>;
  feature_ranges: Record<string, { min: number; max: number }>;
  feature_panel_2023: Record<string, Record<string, number>>;
  feature_panel_2024: Record<string, Record<string, number>>;
  actual_2024: Record<string, number>;
  constants: { gwp_usdm: number; base_lr: number; elasticity: number };
}

const API = import.meta.env.VITE_PIPELINE_API as string | undefined;

let metaCache: PipelineMeta | null = null;
let lastApiError: string | null = null;

// FALLBACK_META is the real meta.json snapshot bundled at build time. All 10 SEA
// countries, 10 features, and the same NGFS scenarios as the server. Eliminates
// the 3-country mislabel risk when API is offline.
const FALLBACK_META: PipelineMeta = metaFile as unknown as PipelineMeta;

export function getLastApiError(): string | null {
  return lastApiError;
}

/**
 * Fetch pipeline metadata. Memoised; falls back to FALLBACK_META if API absent.
 */
export async function getMeta(): Promise<PipelineMeta> {
  if (metaCache) return metaCache;
  if (!API) {
    metaCache = FALLBACK_META;
    return metaCache;
  }
  try {
    const ctrl = new AbortController();
    const t = setTimeout(() => ctrl.abort(), 4000);
    const res = await fetch(`${API}/meta`, { signal: ctrl.signal });
    clearTimeout(t);
    if (!res.ok) throw new Error(`meta ${res.status}`);
    metaCache = (await res.json()) as PipelineMeta;
    lastApiError = null;
    return metaCache;
  } catch (err) {
    lastApiError = err instanceof Error ? err.message : String(err);
    metaCache = FALLBACK_META;
    return metaCache;
  }
}

/**
 * Predict — calls FastAPI; on failure, returns a synthesised cached trace so the UI
 * never sits on an empty state. Pass an external AbortSignal to cancel in-flight
 * requests when a newer one supersedes them (slider drag).
 */
export async function predict(req: PredictRequest, signal?: AbortSignal): Promise<Trace> {
  if (API) {
    try {
      const ctrl = new AbortController();
      const t = setTimeout(() => ctrl.abort(), 4000);
      // Forward external cancellation into our own controller.
      const onAbort = () => ctrl.abort();
      signal?.addEventListener('abort', onAbort);
      const res = await fetch(`${API}/predict`, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(req),
        signal: ctrl.signal,
      });
      clearTimeout(t);
      signal?.removeEventListener('abort', onAbort);
      if (!res.ok) throw new Error(`predict ${res.status}`);
      const trace = (await res.json()) as Trace;
      // Light runtime guard — bad numerics from a hostile or buggy API shouldn't crash the UI.
      if (!Number.isFinite(trace?.stage_3_xgb?.ghg_pred_Mt)) {
        throw new Error('predict response had non-finite ghg_pred_Mt');
      }
      lastApiError = null;
      return trace;
    } catch (err) {
      // External-abort: don't fall back; let the caller handle it.
      if (signal?.aborted) throw err;
      lastApiError = err instanceof Error ? err.message : String(err);
      // fall through to synthesiser
    }
  }
  return synthesise(req);
}

/**
 * Cached-mode synthesiser. Uses canon JSON outputs + the real meta panel to produce
 * a stage-shaped trace. Honest about being cached — `inference_ms` is null and
 * `served_by` reads 'cached'. Forward-mode anchor mirrors serve/pipeline.py:122
 * (compound from M3a-predicted 2024 emissions, NOT from exp(log_GHG_lag1)).
 */
function synthesise(req: PredictRequest): Trace {
  const meta = metaCache ?? FALLBACK_META;
  const targetYear = req.target_year ?? (req.mode === 'hindcast' ? 2024 : 2030);
  const country = meta.feature_panel_2024[req.country] ? req.country : 'Vietnam';
  const baseFeatures = meta.feature_panel_2024[country];
  const overridesIn = req.feature_overrides ?? {};
  const overrides: Record<string, number> = Object.fromEntries(
    Object.entries(overridesIn).filter(([, v]) => v != null && Number.isFinite(v)) as [string, number][],
  );
  const raw: Record<string, number> = { ...baseFeatures, ...overrides };

  const X = meta.m3a_features.map((f) => raw[f] ?? 0);

  // Anchor on canon M3a 2024 prediction for both modes — server does the same.
  // Forward mode then compounds via NGFS scenario from that 2024 anchor.
  const canonM3a = canon.m3a_per_country.find((r) => r.country === country);
  const ghg_pred_Mt = canonM3a?.pred_2024 ?? Math.exp(raw.log_GHG_lag1 ?? 6.3);
  const log_ghg_pred = Math.log(Math.max(ghg_pred_Mt, 0.0001));
  const actual_Mt = req.mode === 'hindcast'
    ? (meta.actual_2024[country] ?? canonM3a?.actual_2024 ?? null)
    : null;
  const err_pct = actual_Mt !== null
    ? ((ghg_pred_Mt - actual_Mt) / actual_Mt) * 100
    : null;

  // Stage 4 — scenario compound growth from 2024 anchor. Mirrors server logic.
  const scenario = req.scenario ?? 'Current Policies';
  const g = meta.ngfs_scenarios[scenario] ?? 0;
  const yearsCompounded = Math.max(0, targetYear - 2024);
  const target_Mt = ghg_pred_Mt * Math.pow(1 + g, yearsCompounded);
  const hothouseG = meta.ngfs_scenarios['Current Policies'] ?? 0.025;
  const hothouse_Mt = ghg_pred_Mt * Math.pow(1 + hothouseG, yearsCompounded);
  const delta_pct = hothouse_Mt > 0 ? (target_Mt - hothouse_Mt) / hothouse_Mt : 0;

  // Stage 5 — loss-ratio mapping.
  const elasticity = req.elasticity ?? meta.constants.elasticity;
  const gwp = req.gwp_usdm ?? meta.constants.gwp_usdm;
  const baseLr = req.base_lr ?? meta.constants.base_lr;
  const lr = baseLr * (1 + elasticity * delta_pct);
  const loss_USDm = gwp * lr;
  const hothouseLr = baseLr; // delta=0 by definition
  const hothouseLoss = gwp * hothouseLr;
  const loss_swing_vs_hothouse_USDm = loss_USDm - hothouseLoss;

  return {
    stage_1_inputs: {
      country, year: targetYear, mode: req.mode,
      raw_features: raw, applied_overrides: overrides,
    },
    stage_2_features: {
      feature_order: meta.m3a_features,
      X,
      log_transformed_keys: meta.m3a_features.filter((f) => f.startsWith('log_')),
    },
    stage_3_xgb: {
      log_ghg_pred, ghg_pred_Mt,
      actual_Mt, err_pct,
      inference_ms: null,
      model: 'M3a',
    },
    stage_4_scenario: {
      scenario,
      growth_rate_pa: g,
      years_compounded: yearsCompounded,
      hothouse_Mt, target_Mt, delta_pct,
    },
    stage_5_loss: {
      lr,
      lr_pp_vs_base: (lr - baseLr) * 100,
      loss_USDm,
      loss_swing_vs_hothouse_USDm,
    },
    trace_meta: {
      pipeline_version: meta.pipeline_version,
      seed: meta.random_state,
      total_latency_ms: null,
      served_by: 'cached',
      served_at: new Date().toISOString(),
    },
  };
}

export const HAS_API = !!API;

File: serve/pipeline.py
"""R-Ignite pipeline runtime.

Loads the M3a (forecast) and M3b (structural) XGBoost models exported by
analysis/python/analysis.ipynb cell 42 and exposes a single
`run_pipeline()` function that returns a five-stage trace as JSON. The trace
shape mirrors what the React Pipeline screen expects.
"""
from __future__ import annotations

import json
import math
import time
from datetime import datetime, timezone
from pathlib import Path
from typing import Literal

import numpy as np
from pydantic import BaseModel, Field, field_validator
from xgboost import XGBRegressor

ROOT = Path(__file__).parent / "models"

m3a = XGBRegressor()
m3a.load_model(str(ROOT / "m3a.json"))

m3b = XGBRegressor()
m3b.load_model(str(ROOT / "m3b.json"))

with open(ROOT / "meta.json") as f:
    META = json.load(f)

# Live data overlay — Climate TRACE snapshot from data/external/live/latest/.
# Additive: bundled META["actual_2024"] (WDI vintage) stays authoritative for
# regression tests and notebook reproducibility; live values exposed under
# "actual_2024_live" + freshness fields so the PWA can render a "data as of"
# pill without breaking the seed=2026 invariant.
_LIVE_DIR = Path(__file__).resolve().parents[1] / "data" / "external" / "live" / "latest"
_LIVE_CT = _LIVE_DIR / "climate_trace.json"
if _LIVE_CT.exists():
    try:
        _snap = json.loads(_LIVE_CT.read_text())
        if isinstance(_snap.get("actual_2024_Mt"), dict):
            META["actual_2024_live"] = _snap["actual_2024_Mt"]
            META["data_as_of"] = _snap.get("fetched_at")
            META["data_source"] = _snap.get("source")
            META["data_url"] = _snap.get("url")
    except (json.JSONDecodeError, OSError):
        pass

# Sanity check: model and meta must agree on input arity. Catches the worst
# silent failure mode — re-running cell 42 after editing FEATURES without
# retraining the M3a binary.
_expected_n = m3a.n_features_in_  # type: ignore[attr-defined]
_meta_n = len(META.get("m3a_features", []))
assert _expected_n == _meta_n, (
    f"meta.json/m3a.json drift: m3a expects {_expected_n} features, meta lists {_meta_n}. "
    "Re-run analysis/python/analysis.ipynb cell 42."
)

_ALLOWED_COUNTRIES = tuple(sorted(META.get("feature_panel_2024", {}).keys()))
_ALLOWED_SCENARIOS = tuple(META.get("ngfs_scenarios", {}).keys())
_FEATURE_RANGES = META.get("feature_ranges", {})
_M3A_FEATURE_SET = set(META.get("m3a_features", []))


class PredictRequest(BaseModel):
    mode: Literal["hindcast", "forward"] = "forward"
    country: str
    scenario: str = "Current Policies"
    elasticity: float = Field(default=0.7, ge=0.0, le=2.0)
    gwp_usdm: float = Field(default=1200.0, ge=0)
    base_lr: float = Field(default=0.62, ge=0, le=2)
    # Bounded to the realistic NGFS horizon — prevents pow-loop blowup and NaN/Inf
    # responses from a crafted target_year=10**9 request.
    target_year: int | None = Field(default=None, ge=2024, le=2050)
    # Cardinality cap prevents a million-key payload from saturating the worker.
    feature_overrides: dict[str, float] | None = Field(default=None, max_length=32)

    @field_validator("country")
    @classmethod
    def _valid_country(cls, v: str) -> str:
        if v not in _ALLOWED_COUNTRIES:
            raise ValueError(
                f"unknown country '{v}' — allowed: {', '.join(_ALLOWED_COUNTRIES)}"
            )
        return v

    @field_validator("scenario")
    @classmethod
    def _valid_scenario(cls, v: str) -> str:
        if v not in _ALLOWED_SCENARIOS:
            raise ValueError(
                f"unknown scenario '{v}' — allowed: {', '.join(_ALLOWED_SCENARIOS)}"
            )
        return v


def run_pipeline(req: PredictRequest) -> dict:
    """Execute the full prediction pipeline and return the 5-stage trace."""
    t_start = time.perf_counter()

    # Stage 1 — assemble inputs.
    # M3a's 2024 row already contains lag1 = 2023 GHG and lag2 = 2022 GHG, which
    # is what the trained model expects for *both* the 2024 hindcast and the
    # 2024-anchored forward projection. The semantic difference is only what we
    # do downstream with the prediction.
    target_year = req.target_year or (2024 if req.mode == "hindcast" else 2030)
    country = req.country  # validated by pydantic above
    base_features = dict(META["feature_panel_2024"][country])

    # Sanitise overrides: only known features, only finite values, clamped to
    # the 5/95 percentile bounds advertised in /meta. Unknown keys silently
    # ignored (request still succeeds — friendlier for live demo than a 422).
    overrides_in = req.feature_overrides or {}
    applied_overrides: dict[str, float] = {}
    raw_features = dict(base_features)
    for k, raw_v in overrides_in.items():
        if k not in _M3A_FEATURE_SET or raw_v is None:
            continue
        v = float(raw_v)
        if not math.isfinite(v):
            continue
        rng = _FEATURE_RANGES.get(k)
        if rng:
            v = max(float(rng["min"]), min(float(rng["max"]), v))
        raw_features[k] = v
        applied_overrides[k] = v

    stage_1 = {
        "country": country,
        "year": target_year,
        "mode": req.mode,
        "raw_features": raw_features,
        "applied_overrides": applied_overrides,
    }

    # Stage 2 — feature vector in M3a order
    feature_order: list[str] = META["m3a_features"]
    X = [float(raw_features.get(f, 0.0)) for f in feature_order]
    log_keys = [k for k in feature_order if k.startswith("log_")]

    stage_2 = {
        "feature_order": feature_order,
        "X": X,
        "log_transformed_keys": log_keys,
    }

    # Stage 3 — XGBoost inference
    X_arr = np.asarray([X], dtype=np.float32)
    t_inf = time.perf_counter()
    log_ghg_pred = float(m3a.predict(X_arr)[0])
    inference_ms = (time.perf_counter() - t_inf) * 1000.0
    # Clamp before exp() — defends against extreme override values producing
    # OverflowError. log(1e9 Mt) ≈ 20.7, so 30 is a generous ceiling.
    ghg_pred_Mt = math.exp(min(log_ghg_pred, 30.0))

    actual_Mt = None
    err_pct = None
    if req.mode == "hindcast":
        actual = META.get("actual_2024", {}).get(country)
        if actual is not None:
            actual_Mt = float(actual)
            err_pct = ((ghg_pred_Mt - actual_Mt) / actual_Mt) * 100.0

    stage_3 = {
        "log_ghg_pred": log_ghg_pred,
        "ghg_pred_Mt": ghg_pred_Mt,
        "actual_Mt": actual_Mt,
        "err_pct": err_pct,
        "inference_ms": round(inference_ms, 3),
        "model": "M3a",
    }

    # Stage 4 — NGFS scenario overlay.
    # Compound from the M3a-predicted 2024 emissions to target_year.
    scenarios: dict[str, float] = META["ngfs_scenarios"]
    g = float(scenarios[req.scenario])
    hothouse_g = float(scenarios.get("Current Policies", 0.025))
    years = max(0, target_year - 2024)
    target_Mt = ghg_pred_Mt * (1.0 + g) ** years
    hothouse_Mt = ghg_pred_Mt * (1.0 + hothouse_g) ** years
    delta_pct = (target_Mt - hothouse_Mt) / hothouse_Mt if hothouse_Mt > 0 else 0.0

    stage_4 = {
        "scenario": req.scenario,
        "growth_rate_pa": g,
        "years_compounded": years,
        "hothouse_Mt": hothouse_Mt,
        "target_Mt": target_Mt,
        "delta_pct": delta_pct,
    }

    # Stage 5 — loss-ratio mapping
    lr = req.base_lr * (1.0 + req.elasticity * delta_pct)
    loss_USDm = req.gwp_usdm * lr
    hothouse_lr = req.base_lr  # at delta=0
    hothouse_loss = req.gwp_usdm * hothouse_lr
    loss_swing = loss_USDm - hothouse_loss

    stage_5 = {
        "lr": lr,
        "lr_pp_vs_base": (lr - req.base_lr) * 100.0,
        "loss_USDm": loss_USDm,
        "loss_swing_vs_hothouse_USDm": loss_swing,
    }

    total_latency_ms = (time.perf_counter() - t_start) * 1000.0

    return {
        "stage_1_inputs": stage_1,
        "stage_2_features": stage_2,
        "stage_3_xgb": stage_3,
        "stage_4_scenario": stage_4,
        "stage_5_loss": stage_5,
        "trace_meta": {
            "pipeline_version": META.get("pipeline_version", "1.0"),
            "seed": META.get("random_state", 2026),
            "total_latency_ms": round(total_latency_ms, 3),
            "served_by": "fastapi",
            "served_at": datetime.now(timezone.utc).isoformat(),
        },
    }
